File: assignment-a/src/yarn.ts
/**
 * Yarn quantity calculator.
 *
 * Model (all steps explicit and easy to change — see assumptions.md):
 *
 *   metres = area_cm2 * metresPerCm2(weight) * stitchFactor * gaugeFactor
 *
 * where
 *   area_cm2      = widthCm * heightCm
 *   metresPerCm2  = per-weight constant (finer yarn -> more metres per area)
 *   stitchFactor  = relative usage of the stitch pattern vs stockinette
 *   gaugeFactor   = userGauge / typicalGauge (knitting tighter uses more yarn)
 *
 * Balls/skeins = ceil( metresWithSafety / metresPerBall ).
 */

import { CalcError, resolveWeight, resolveStitchFactor, roundTo } from "./data.js";

export interface YarnEstimate {
  metres: number;
  metresWithSafety: number;
  balls: number;
  metresPerBall: number;
  inputs: {
    widthCm: number;
    heightCm: number;
    weight: string;
    weightLabel: string;
    stitch: string;
    gaugeStsPer10cm: number | null;
    safety: number;
  };
  assumptions: {
    metresPerCm2: number;
    stitchFactor: number;
    gaugeFactor: number;
    typicalGaugeStsPer10cm: number;
  };
}

export interface YarnOptions {
  stitch?: string;
  gaugeStsPer10cm?: number | null;
  metresPerBall?: number | null;
  safety?: number;
}

/**
 * Estimate yarn required for a flat rectangular piece.
 * Throws CalcError on invalid inputs so the assistant can decline cleanly.
 */
export function estimateYarn(
  widthCm: number,
  heightCm: number,
  weight: string,
  options: YarnOptions = {},
): YarnEstimate {
  const { stitch = "stockinette", gaugeStsPer10cm = null, metresPerBall = null, safety = 0.1 } = options;

  if (widthCm === null || widthCm === undefined || heightCm === null || heightCm === undefined) {
    throw new CalcError("Both widthCm and heightCm are required.");
  }
  if (widthCm <= 0 || heightCm <= 0) {
    throw new CalcError("Dimensions must be positive.");
  }
  if (safety < 0) {
    throw new CalcError("Safety margin cannot be negative.");
  }

  const w = resolveWeight(weight);
  const [stitchName, stitchFactor] = resolveStitchFactor(stitch);

  let gaugeFactor = 1.0;
  if (gaugeStsPer10cm !== null && gaugeStsPer10cm !== undefined) {
    if (gaugeStsPer10cm <= 0) throw new CalcError("Gauge must be positive.");
    gaugeFactor = gaugeStsPer10cm / w.typicalGauge;
  }

  const area = widthCm * heightCm;
  const metres = area * w.metresPerCm2 * stitchFactor * gaugeFactor;
  const metresWithSafety = metres * (1 + safety);

  const mpb = metresPerBall ?? w.metresPerBall;
  if (mpb <= 0) throw new CalcError("metresPerBall must be positive.");
  const balls = Math.ceil(metresWithSafety / mpb);

  return {
    metres: Math.round(metres),
    metresWithSafety: Math.round(metresWithSafety),
    balls,
    metresPerBall: mpb,
    inputs: {
      widthCm,
      heightCm,
      weight: w.slug,
      weightLabel: w.label,
      stitch: stitchName,
      gaugeStsPer10cm,
      safety,
    },
    assumptions: {
      metresPerCm2: w.metresPerCm2,
      stitchFactor,
      gaugeFactor: roundTo(gaugeFactor, 3),
      typicalGaugeStsPer10cm: w.typicalGauge,
    },
  };
}

export function formatYarn(e: YarnEstimate): string {
  const i = e.inputs;
  return (
    `For a ${i.widthCm} x ${i.heightCm} cm ${i.stitch} piece in ${i.weightLabel} yarn, ` +
    `you need about ${e.metres} m of yarn (~${e.metresWithSafety} m with a ` +
    `${Math.round(i.safety * 100)}% safety margin), which is about ${e.balls} ball(s) ` +
    `at ~${e.metresPerBall} m per ball.`
  );
}

File: assignment-a/knit_calc/yarn.py
"""Yarn quantity calculator.

Model (all steps explicit and easy to change — see assumptions.md):

    metres = area_cm2 * metres_per_cm2(weight) * stitch_factor * gauge_factor

where
    area_cm2       = width_cm * height_cm
    metres_per_cm2 = per-weight constant (finer yarn -> more metres per area)
    stitch_factor  = relative usage of the stitch pattern vs stockinette
    gauge_factor   = user_gauge / typical_gauge  (knitting tighter uses more yarn)

Balls/skeins = ceil( metres_with_safety / metres_per_ball ).
"""

import math
from dataclasses import dataclass, field, asdict
from typing import Optional

from .data import resolve_weight, resolve_stitch_factor


@dataclass
class YarnEstimate:
    metres: float
    metres_with_safety: float
    balls: int
    metres_per_ball: float
    inputs: dict = field(default_factory=dict)
    assumptions: dict = field(default_factory=dict)

    def to_dict(self) -> dict:
        return asdict(self)

    def summary(self) -> str:
        i = self.inputs
        return (
            f"For a {i['width_cm']} x {i['height_cm']} cm {i['stitch']} piece in "
            f"{i['weight_label']} yarn, you need about {self.metres} m of yarn "
            f"(~{self.metres_with_safety} m with a {int(i['safety']*100)}% safety margin), "
            f"which is about {self.balls} ball(s) at ~{self.metres_per_ball} m per ball."
        )


def estimate_yarn(
    width_cm: float,
    height_cm: float,
    weight: str,
    stitch: str = "stockinette",
    gauge_sts_per_10cm: Optional[float] = None,
    metres_per_ball: Optional[float] = None,
    safety: float = 0.10,
) -> YarnEstimate:
    """Estimate yarn required for a flat rectangular piece.

    Raises ValueError on invalid inputs so the assistant can decline cleanly.
    """
    if width_cm is None or height_cm is None:
        raise ValueError("Both width_cm and height_cm are required.")
    if width_cm <= 0 or height_cm <= 0:
        raise ValueError("Dimensions must be positive.")
    if safety < 0:
        raise ValueError("Safety margin cannot be negative.")

    w = resolve_weight(weight)
    stitch_name, stitch_factor = resolve_stitch_factor(stitch)

    gauge_factor = 1.0
    if gauge_sts_per_10cm is not None:
        if gauge_sts_per_10cm <= 0:
            raise ValueError("Gauge must be positive.")
        gauge_factor = gauge_sts_per_10cm / w.typical_gauge

    area = width_cm * height_cm
    metres = area * w.metres_per_cm2 * stitch_factor * gauge_factor
    metres_with_safety = metres * (1 + safety)

    mpb = metres_per_ball if metres_per_ball is not None else w.metres_per_ball
    if mpb <= 0:
        raise ValueError("metres_per_ball must be positive.")
    balls = math.ceil(metres_with_safety / mpb)

    return YarnEstimate(
        metres=round(metres),
        metres_with_safety=round(metres_with_safety),
        balls=balls,
        metres_per_ball=mpb,
        inputs={
            "width_cm": width_cm,
            "height_cm": height_cm,
            "weight": w.slug,
            "weight_label": w.label,
            "stitch": stitch_name,
            "gauge_sts_per_10cm": gauge_sts_per_10cm,
            "safety": safety,
        },
        assumptions={
            "metres_per_cm2": w.metres_per_cm2,
            "stitch_factor": stitch_factor,
            "gauge_factor": round(gauge_factor, 3),
            "typical_gauge_sts_per_10cm": w.typical_gauge,
        },
    )
